mountpoint-s3/src/upload.rs

```diff
@@ -0,0 +1,101 @@
+use std::sync::Arc;
+
+use mountpoint_s3_client::{
+    ObjectClient, ObjectClientError, ObjectClientResult, PutObjectError, PutObjectParams, PutObjectRequest,
+    PutObjectResult,
+};
+
+use thiserror::Error;
+
+type PutRequestError<Client> = ObjectClientError<PutObjectError, <Client as ObjectClient>::ClientError>;
+
+/// An [Uploader] creates and manages streaming PutObject requests.
```

Comment:
Can you explain a little more what you picture the role of this thing being beyond a `PutObjectRequest`? If I'm reading it right, the big thing it does is enforce sequential writes (which the current write code inlines into the `write` FUSE handler). Will it do other things too, like checksums? Does it do its own buffering?

Reply:
Yes, that's all it is now. It could be a good place to put additional logic that is specific to mountpoint (vs generic for PutObjectRequest).
Checksums, buffering: both are likely candidates, but I do not have a clear picture yet. For example, if we decide to support some form of local staging (e.g. implement fallocate), I would put all that logic here.